File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LockOnThisOrType.cs
using System;
using System.Collections.Generic;

namespace Tests.Diagnostics
{
    public class LockOnThisOrType
    {
        public void MyLockingMethod()
        {
            lock (this) // Noncompliant
//                ^^^^
            {
                // ...
            }

            lock (lockObj)
            {
                // ...
            }

            lock (typeof(LockOnThisOrType)) // Noncompliant
//                ^^^^^^^^^^^^^^^^^^^^^^^^
            {
                // ...
            }
            lock ((new LockOnThisOrType()).GetType()) // Noncompliant {{Lock on a dedicated object instance instead.}}
            {
                // ...
            }
        }

        object lockObj = new object();
    }
}

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LockOnThisOrType.vb
Imports System
Imports System.Collections.Generic

Namespace Tests.Diagnostics
    Public Class LockOnThisOrType
        Public Sub MyLockingMethod()
            SyncLock Me ' Noncompliant {{Lock on a dedicated object instance instead.}}
'                    ^^
            End SyncLock

            SyncLock lockObj
            End SyncLock

            SyncLock GetType(LockOnThisOrType) ' Noncompliant
'                    ^^^^^^^^^^^^^^^^^^^^^^^^^
            End SyncLock

            SyncLock (New LockOnThisOrType()).[GetType]() ' Noncompliant
'                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            End SyncLock
        End Sub

        Dim lockObj As New Object()
    End Class
End Namespace
